File: app/src/main/java/it/cammino/risuscito/CambioAccordi.java
package it.cammino.risuscito;

import android.content.Context;
import android.util.Log;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import it.cammino.risuscito.ui.ThemeableActivity;

public class CambioAccordi {

    private static final String TAG = "CambioAccordi";

    static final String[] accordi_it =
            {"Do", "Do#", "Re", "Mib", "Mi", "Fa", "Fa#", "Sol", "Sol#", "La", "Sib", "Si"};

    static final String[] accordi_uk =
            {"C", "Cis", "D","Eb", "E", "F", "Fis", "G", "Gis", "A", "B","H"};

    private static final String[] accordi_uk_lower =
            {"c", "cis", "d","eb", "e", "f", "fis", "g", "gis", "a", "b","h"};

    static final String[] accordi_en =
            {"C", "C#", "D","Eb", "E", "F", "F#", "G", "G#", "A", "Bb","B"};

    static final String[] barre_it = {"I", "II", "III", "IV", "V", "VI", "VII"};
    static final String[] barre_uk = {"I", "II", "III", "IV", "V", "VI", "VII"};
    static final String[] barre_en = {"1", "2", "3", "4", "5", "6", "7"};

    private Context context;
    private String mLanguage;

    CambioAccordi(Context context) {
        this.context = context;
    }

    public CambioAccordi(Context context, String language) {
        this.context = context;
        this.mLanguage = language;
    }

    static String recuperaPrimoAccordo(InputStream canto, String language) {

        if (canto == null)
            return "";

        StringBuilder primaNota = new StringBuilder();

        try {
            BufferedReader br = new BufferedReader(
                    new InputStreamReader(
                            canto, "UTF-8"));

            String line = br.readLine();
            boolean found = false;

            while (line != null && !found) {
                if (line.contains("A13F3C") && !line.contains("<H2>") && !line.contains("<H4>")) {
                    Log.v(TAG, "recuperaPrimoAccordo - RIGA: " + line);
                    int inizioRiga = line.indexOf("A13F3C") + 8;

                    if (inizioRiga < line.length()) {
                        Log.v(TAG, "recuperaPrimoAccordo - inizioRiga: " + inizioRiga);
                        Log.v(TAG, "recuperaPrimoAccordo - carattere: " + line.charAt(inizioRiga));
                        int i = inizioRiga;
                        while (i < line.length()) {
                            Log.v(TAG, "recuperaPrimoAccordo - LETTERA: " + line.charAt(i));
                            if (line.charAt(i) != ' ') {
                                found = true;
                                break;
                            }
                            i++;
                        }
                        Log.v(TAG, "recuperaPrimoAccordo - inizio Nota: " + i);
                        Log.v(TAG, "recuperaPrimoAccordo - lunghezza stringa: " + line.length());
                        primaNota.append(line.charAt(i));
                        Log.v(TAG, "recuperaPrimoAccordo - prima lettera: " + primaNota);
                        for (int j = i+1; j < line.length(); j++) {
                            Log.v(TAG, "recuperaPrimoAccordo - DA ISP: " + line.charAt(j));
//                            Matcher myMatcher = Pattern.compile("[a-z]|#]")
//                                    .matcher(String.valueOf(line.charAt(j)));
//                            if (language.equalsIgnoreCase("en"))
//                                myMatcher = Pattern.compile("[^m][a-z]|#]")
//                                        .matcher(String.valueOf(line.charAt(j)));
                            Matcher myMatcher;
                            switch (language) {
                                case "en":
                                    myMatcher = Pattern.compile("[^m][a-z]|#]")
                                            .matcher(String.valueOf(line.charAt(j)));
                                    break;
                                default:
                                    myMatcher = Pattern.compile("[a-z]|#]")
                                            .matcher(String.valueOf(line.charAt(j)));
                                    break;
                            }
                            if (myMatcher.find()) {
                                Log.v(TAG, "recuperaPrimoAccordo - matchato OK");
                                primaNota.append(line.charAt(j));
                            }
                            else
                                break;
                        }
                    }

                }
                line = br.readLine();
            }
            br.close();
            Log.v(TAG, "recuperaPrimoAccordo - risultato: " + primaNota);
            return primaNota.toString();
        }
        catch (Exception ex)
        {
            ex.printStackTrace();
            return "";
        }
    }

    String recuperaBarre(InputStream canto, String language) {

        if (canto == null)
            return "";

//        String language = context.getResources().getConfiguration().locale.getLanguage();

        String primoBarre = "0";

        try {
            BufferedReader br = new BufferedReader(
                    new InputStreamReader(
                            canto, "UTF-8"));

            String line = br.readLine();
            boolean found = false;

            while (line != null && !found) {
                int start = line.indexOf(context.getResources().getString(R.string.barre_search_string));
//                if (line.contains(context.getResources().getString(R.string.barre_search_string))) {
                if (start >= 0) {
                    Log.v(TAG, "recuperaBarre - RIGA: " + line);
                    found = true;

//                    int start = line.indexOf(context.getResources().getString(R.string.barre_add_al)) + 3;
                    start = language.equalsIgnoreCase("en")? start + 5 : (line.indexOf(context.getResources().getString(R.string.barre_add_al)) + 3);

                    StringBuilder primoBarreBuilder = new StringBuilder();
                    for (int i = start; i < line.length(); i++) {
                        if (line.charAt(i) == ' ' || line.charAt(i) == '<')
                            break;
                        else
                            primoBarreBuilder.append(line.charAt(i));
                    }
                    primoBarre = primoBarreBuilder.toString();
                }
                line = br.readLine();
            }
            br.close();
            Log.v(TAG, "recuperaBarre - risultato: " + primoBarre);
            return primoBarre;
        }
        catch (Exception ex) {
            Log.e(getClass().getName(), ex.getLocalizedMessage(), ex);
            return "";
        }
    }

    public HashMap<String, String> diffSemiToni(String primaNota, String notaCambio) {

//		if (primaNota.equals(notaCambio))
        Log.v(TAG, "diffSemiToni - primaNota: " + primaNota);
        Log.v(TAG, "diffSemiToni - notaCambio: " + notaCambio);
//			return null;

        if (primaNota == null || primaNota.equals("")
                || notaCambio == null || notaCambio.equals(""))
            return null;

//        String language = context.getResources().getConfiguration().locale.getLanguage();
        String language = ThemeableActivity.getSystemLocalWrapper(context.getResources().getConfiguration()).getLanguage();
        if (mLanguage != null && !mLanguage.isEmpty())
            language = mLanguage;

        Log.v(TAG, "diffSemiToni: language " + language);

        String primoAccordo = primaNota;
        String cambioAccordo = notaCambio;

        String[] accordi;
//        String[] accordi = accordi_it;
//        if (language.equalsIgnoreCase("uk")) {
//            accordi = accordi_uk;
//            if (primoAccordo.length() == 1)
//                primoAccordo = primoAccordo.toUpperCase();
//            else
//                primoAccordo = primoAccordo.substring(0,1).toUpperCase() + primoAccordo.substring(1);
//            if (cambioAccordo.length() == 1)
//                cambioAccordo = cambioAccordo.toUpperCase();
//            else
//                cambioAccordo = cambioAccordo.substring(0,1).toUpperCase() + cambioAccordo.substring(1);
//        }
//
//        if (language.equalsIgnoreCase("en"))
//            accordi = accordi_en;

        switch (language) {
            case "it":
                accordi = accordi_it;
                break;
            case "uk":
                accordi = accordi_uk;
                if (primoAccordo.length() == 1)
                    primoAccordo = primoAccordo.toUpperCase();
                else
                    primoAccordo = primoAccordo.substring(0,1).toUpperCase() + primoAccordo.substring(1);
                if (cambioAccordo.length() == 1)
                    cambioAccordo = cambioAccordo.toUpperCase();
                else
                    cambioAccordo = cambioAccordo.substring(0,1).toUpperCase() + cambioAccordo.substring(1);
                break;
            case "en":
                accordi = accordi_en;
                break;
            default:
                accordi = accordi_it;
                break;
        }

        int start;
        for (start = 0; start < accordi.length; start++) {
            if (primoAccordo.equals(accordi[start]))
                break;
        }
        if (start == accordi.length)
            return null;
        Log.v(TAG, "diffSemiToni - posizionePrimaNota: " + start);
        int end;
        for (end = 0; end < accordi.length; end++) {
            if (cambioAccordo.equals(accordi[end]))
                break;
        }
        if (end == accordi.length)
            return null;
        Log.v(TAG, "diffSemiToni - posizioneNotaCambio: " + end);
        int differenza;
        if (end > start)
            differenza = (end - start);
        else
            differenza = (end + 12 - start);

        HashMap<String, String> mappa = new HashMap<>();
        for (int i = 0; i < accordi.length; i++) {
            Log.v(TAG, "diffSemiToni - NUOVO: " + (i+differenza)%12);
            Log.v(TAG, "diffSemiToni - CONVE: " + accordi[i] + " in " + accordi[(i+differenza)%12]);
            mappa.put(accordi[i], accordi[(i+differenza)%12]);
        }
        return mappa;
    }

    public HashMap<String, String> diffSemiToniMin(String primaNota, String notaCambio) {

        Log.v(TAG, "diffSemiToniMin");

        if (primaNota == null || primaNota.equals("")
                || notaCambio == null || primaNota.equals(""))
            return null;

        String primoAccordo = primaNota;
        if (primoAccordo.length() == 1)
            primoAccordo = primoAccordo.toLowerCase();
        else
            primoAccordo = primoAccordo.substring(0,1).toLowerCase() + primoAccordo.substring(1);
        String cambioAccordo = notaCambio;
        if (cambioAccordo.length() == 1)
            cambioAccordo = cambioAccordo.toLowerCase();
        else
            cambioAccordo = cambioAccordo.substring(0,1).toLowerCase() + cambioAccordo.substring(1);

        int start;
        for (start = 0; start < accordi_uk_lower.length; start++) {
            if (primoAccordo.equals(accordi_uk_lower[start]))
                break;
        }
        if (start == accordi_uk_lower.length)
            return null;
        Log.v(TAG, "diffSemiToniMin - posizionePrimaNota: " + start);
        int end;
        for (end = 0; end < accordi_uk_lower.length; end++) {
            if (cambioAccordo.equals(accordi_uk_lower[end]))
                break;
        }
        if (end == accordi_uk_lower.length)
            return null;
        Log.v(TAG, "diffSemiToniMin - posizioneNotaCambio: " + end);
        int differenza;
        if (end > start)
            differenza = (end - start);
        else
            differenza = (end + 12 - start);

        HashMap<String, String> mappa = new HashMap<>();
        for (int i = 0; i < accordi_uk_lower.length; i++) {
            Log.v(TAG, "diffSemiToniMin - NUOVO: " + (i+differenza)%12);
            Log.v(TAG, "diffSemiToniMin - CONVE: " + accordi_uk_lower[i] + " in " + accordi_uk_lower[(i + differenza) % 12]);
            mappa.put(accordi_uk_lower[i], accordi_uk_lower[(i+differenza)%12]);
        }
        return mappa;
    }
}

File: app/src/main/java/it/cammino/risuscito/CambioAccordi.kt
package it.cammino.risuscito

import android.content.Context
import android.util.Log
import com.google.firebase.crashlytics.FirebaseCrashlytics
import it.cammino.risuscito.ui.LocaleManager.Companion.LANGUAGE_ENGLISH
import it.cammino.risuscito.ui.LocaleManager.Companion.LANGUAGE_ITALIAN
import it.cammino.risuscito.ui.LocaleManager.Companion.LANGUAGE_POLISH
import it.cammino.risuscito.ui.LocaleManager.Companion.LANGUAGE_TURKISH
import it.cammino.risuscito.ui.LocaleManager.Companion.LANGUAGE_UKRAINIAN
import it.cammino.risuscito.ui.LocaleManager.Companion.getSystemLocale
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.util.*
import java.util.regex.Pattern

class CambioAccordi internal constructor(private val mContext: Context) {

    internal fun recuperaBarre(canto: InputStream?, language: String): String {

        if (canto == null) return ""

        var primoBarre = "0"

        try {
            val br = BufferedReader(InputStreamReader(canto, "UTF-8"))

            var line: String? = br.readLine()
            var found = false

            while (line != null && !found) {
                var start = line.indexOf(mContext.resources.getString(R.string.barre_search_string))
                if (start >= 0) {
                    Log.v(TAG, "recuperaBarre - RIGA: $line")
                    found = true

                    start = when (language) {
                        LANGUAGE_ENGLISH, LANGUAGE_TURKISH -> start + 5
                        LANGUAGE_POLISH -> start + 6
                        else -> line.indexOf(mContext.resources.getString(R.string.barre_add_al)) + 3
                    }

                    val primoBarreBuilder = StringBuilder()
                    for (i in start until line.length) {
                        if (line[i] == ' ' || line[i] == '<')
                            break
                        else
                            primoBarreBuilder.append(line[i])
                    }
                    primoBarre = primoBarreBuilder.toString()
                }
                line = br.readLine()
            }
            br.close()
            Log.v(TAG, "recuperaBarre - risultato: $primoBarre")
            return primoBarre
        } catch (ex: Exception) {
            Log.e(javaClass.name, ex.localizedMessage, ex)
            return ""
        }

    }

    fun diffSemiToni(primaNota: String?, notaCambio: String?): HashMap<String, String>? {

        primaNota ?: return null
        notaCambio ?: return null

        Log.v(TAG, "diffSemiToni - primaNota: $primaNota")
        Log.v(TAG, "diffSemiToni - notaCambio: $notaCambio")

        if (primaNota == "" || notaCambio == "")
            return null

        val language = getSystemLocale(mContext.resources).language
        val locale = getSystemLocale(mContext.resources)

        Log.v(TAG, "diffSemiToni: language $language")

        var primoAccordo: String = primaNota
        var cambioAccordo: String = notaCambio

        val accordi: Array<String> =
                when (language) {
                    LANGUAGE_ITALIAN, LANGUAGE_TURKISH -> accordi_it
                    LANGUAGE_UKRAINIAN -> accordi_uk
                    LANGUAGE_POLISH -> accordi_pl
                    LANGUAGE_ENGLISH -> accordi_en
//            LANGUAGE_TURKISH -> accordi = accordi_tr
                    else -> accordi_it
                }

        if (language == LANGUAGE_UKRAINIAN || language == LANGUAGE_POLISH) {
            primoAccordo = if (primoAccordo.length == 1)
                primoAccordo.toUpperCase(locale)
            else
                primoAccordo.substring(0, 1).toUpperCase(locale) + primoAccordo.substring(1)
            cambioAccordo = if (cambioAccordo.length == 1)
                cambioAccordo.toUpperCase(locale)
            else
                cambioAccordo.substring(0, 1).toUpperCase(locale) + cambioAccordo.substring(1)
        }

        var start = 0
        while (start < accordi.size) {
            if (primoAccordo == accordi[start]) break
            start++
        }
        if (start == accordi.size) return null
        Log.v(TAG, "diffSemiToni - posizionePrimaNota: $start")
        var end = 0
        while (end < accordi.size) {
            if (cambioAccordo == accordi[end]) break
            end++
        }
        if (end == accordi.size) return null
        Log.v(TAG, "diffSemiToni - posizioneNotaCambio: $end")

        val differenza = if (end > start)
            end - start
        else
            end + 12 - start

        val mappa = HashMap<String, String>()
        for (i in accordi.indices) {
            Log.v(TAG, "diffSemiToni - NUOVO: " + (i + differenza) % 12)
            Log.v(TAG, "diffSemiToni - CONVE: " + accordi[i] + " in " + accordi[(i + differenza) % 12])
            mappa[accordi[i]] = accordi[(i + differenza) % 12]
        }
        return mappa
    }

    fun diffSemiToniMin(primaNota: String?, notaCambio: String?): HashMap<String, String>? {

        val language = getSystemLocale(mContext.resources).language
        val locale = getSystemLocale(mContext.resources)

        Log.v(TAG, "diffSemiToniMin - language: $language")

        if (primaNota == null || primaNota == "" || notaCambio == null || primaNota == "")
            return null

        var primoAccordo: String = primaNota
        primoAccordo = if (primoAccordo.length == 1)
            primoAccordo.toLowerCase(locale)
        else
            primoAccordo.substring(0, 1).toLowerCase(locale) + primoAccordo.substring(1)
        var cambioAccordo: String = notaCambio
        cambioAccordo = if (cambioAccordo.length == 1)
            cambioAccordo.toLowerCase(locale)
        else
            cambioAccordo.substring(0, 1).toLowerCase(locale) + cambioAccordo.substring(1)

        val accordiMin: Array<String> =
                when (language) {
                    LANGUAGE_UKRAINIAN -> accordi_uk_lower
                    LANGUAGE_POLISH -> accordi_pl_lower
                    else -> accordi_uk_lower
                }

        var start = 0
        while (start < accordiMin.size) {
            if (primoAccordo == accordiMin[start]) break
            start++
        }
        if (start == accordiMin.size) return null
        Log.v(TAG, "diffSemiToniMin - posizionePrimaNota: $start")
        var end = 0
        while (end < accordiMin.size) {
            if (cambioAccordo == accordiMin[end]) break
            end++
        }
        if (end == accordiMin.size) return null
        Log.v(TAG, "diffSemiToniMin - posizioneNotaCambio: $end")
        val differenza = if (end > start)
            end - start
        else
            end + 12 - start

        val mappa = HashMap<String, String>()
        for (i in accordiMin.indices) {
            Log.v(TAG, "diffSemiToniMin - NUOVO: " + (i + differenza) % 12)
            Log.v(
                    TAG,
                    "diffSemiToniMin - CONVE: "
                            + accordiMin[i]
                            + " in "
                            + accordiMin[(i + differenza) % 12])
            mappa[accordiMin[i]] = accordiMin[(i + differenza) % 12]
        }
        return mappa
    }

    companion object {

        internal val accordi_it = arrayOf("Do", "Do#", "Re", "Mib", "Mi", "Fa", "Fa#", "Sol", "Sol#", "La", "Sib", "Si")
        internal val accordi_uk = arrayOf("C", "Cis", "D", "Eb", "E", "F", "Fis", "G", "Gis", "A", "B", "H")
        internal val accordi_pl = arrayOf("C", "Cis", "D", "Dis", "E", "F", "Fis", "G", "Gis", "A", "B", "H")
        internal val accordi_en = arrayOf("C", "C#", "D", "Eb", "E", "F", "F#", "G", "G#", "A", "Bb", "B")
        internal val barre_it = arrayOf("0", "I", "II", "III", "IV", "V", "VI", "VII")
        internal val barre_en = arrayOf("0", "1", "2", "3", "4", "5", "6", "7")
        private val TAG = CambioAccordi::class.java.canonicalName
        private val accordi_uk_lower = arrayOf("c", "cis", "d", "eb", "e", "f", "fis", "g", "gis", "a", "b", "h")
        private val accordi_pl_lower = arrayOf("c", "cis", "d", "dis", "e", "f", "fis", "g", "gis", "a", "b", "h")

        internal fun recuperaPrimoAccordo(canto: InputStream?, language: String): String {

            if (canto == null) return ""

            val primaNota = StringBuilder()

            try {
                val br = BufferedReader(InputStreamReader(canto, "UTF-8"))

                var line: String? = br.readLine()
                var found = false

                while (line != null && !found) {
                    if (line.contains("A13F3C") && !line.contains("<H2>") && !line.contains("<H4>")) {
                        Log.v(TAG, "recuperaPrimoAccordo - RIGA: $line")
                        val inizioRiga = line.indexOf("A13F3C") + 8

                        if (inizioRiga < line.length) {
                            Log.v(TAG, """recuperaPrimoAccordo - inizioRiga: $inizioRiga""")
                            Log.v(TAG, "recuperaPrimoAccordo - carattere: " + line[inizioRiga])
                            var i = inizioRiga
                            while (i < line.length) {
                                Log.v(TAG, "recuperaPrimoAccordo - LETTERA: " + line[i])
                                if (line[i] != ' ') {
                                    found = true
                                    break
                                }
                                i++
                            }
                            Log.v(TAG, "recuperaPrimoAccordo - inizio Nota: $i")
                            Log.v(TAG, "recuperaPrimoAccordo - lunghezza stringa: " + line.length)
                            primaNota.append(line[i])
                            Log.v(TAG, "recuperaPrimoAccordo - prima lettera: $primaNota")
                            for (j in i + 1 until line.length) {
                                Log.v(TAG, "recuperaPrimoAccordo - DA ISP: " + line[j])
                                val myMatcher = when (language) {
                                    LANGUAGE_ENGLISH -> Pattern.compile("[^m][a-z]|#]").matcher(line[j].toString())
                                    else -> Pattern.compile("[a-z]|#]").matcher(line[j].toString())
                                }
                                if (myMatcher.find()) {
                                    Log.v(TAG, "recuperaPrimoAccordo - matchato OK")
                                    primaNota.append(line[j])
                                } else
                                    break
                            }
                        }
                    }
                    line = br.readLine()
                }
                br.close()
                Log.v(TAG, "recuperaPrimoAccordo - risultato: $primaNota")
                return primaNota.toString()
            } catch (ex: Exception) {
                Log.e(TAG, "Error:", ex)
                FirebaseCrashlytics.getInstance().recordException(ex)
                return ""
            }

        }
    }
}
